Source/Dafny/Verifier/Translator.TrStatement.cs

```diff
@@ -844,6 +844,10 @@ public partial class Translator {
               needDefiniteAssignmentTracking = true;
             }
           }
+          if (!local.Type.IsNonempty) {
+            // This prevents generating an unsatisfiable where clause for possibly empty types
```

Comment:
Thanks.

I'm still a little confused about empty types. If the type doesn't have values, and `GimmieOne` return a value of this type, then I'd expect an error in `GimmieOne`.

Should
```
method GimmieOne<W>(s: seq<W>) returns (c: W)
  requires s != []
{
  return s[0];
}
```
even pass the resolver? Why doesn't it require `W(00)` ?

Reply:
You are absolutely correct that `GimmieOne`'s precondition can be satisfied only if `W` is a nonempty type, and likewise the `if s != []` branch can be taken only if `W` is a nonempty type.

The statement `w: W :| b ==> P(w);` incurs a proof obligation `exists w: W :: b ==> P(w)`. At first, you might think this proof obligation holds trivially if `b` is `false`, but one still needs to prove `exists w: W`. So, this existential quantifier is not provable for an empty type. Therefore, the verifier _should_ give an error for this assignment.

Any statement that comes _after_ the `w :| ...` statement should not have any effect on whether or not you get an error for the `w :| ...` statement. Therefore, an error should be reported for the `w :| ...` statement whether or not the `if` statement is included in the method. But the previous unsoundness was that the `if` statement _did_ cause a change.

How did the presence of the `if ... { var cand := ...; }` statement cause the verifier not to flag `w :| ...` as an error? Previously (before this PR), the Boogie generated for method `M` looked something like this (where I'm omitting many irrelevant details):

```
procedure M(W: Ty, s: Seq Box where ...) returns (defass$r: bool, r: Box where defass$r ==> $Is(r, W))
{
  var b: bool;
  var defass$w: bool;
  var w: Box where defass$w ==> $Is(w, W);
  var $rhs: Box where $Is($rhs, W);
  var cand: Box where $Is($rhs, W);

  b := exists w: Box :: $Is(w, W) && P(w);

  assert exists w: Box :: $Is(w, W) && (b ==> P(w));
  havoc w;
  assume b ==> P(w);
  defass$w := true;

  if s != [] {
    call $rhs := GimmieOne(s);
    cand := $rhs;
  }

  assert defass$w; // check that w has been defined
  r := w;
  defass$r := true;
  assert defass$r; // check that r has been defined before returning
  return;
}
```

Note that out-parameter `r` and local variable `w` have `where` clauses that are guarded by `defass$r` and `defass$w` (respectively), which are set to `true` when `r` and `w` are known to contain a value that satisfies `$Is(_, W)`. Note also that local variables `$rhs` and `cand` (which come about because of the `if` statement downstream of the `w :| ...` assignment) have `where` clauses that are _not_ guarded by a `defass$` variable. Boogie assumes all `where` clauses of in- and out-parameters and local variables on entry to the procedure body (and in all places where Boogie gets to assign an arbitrary value to these variables; see Boogie reference manual for details). So, because of the `where` clauses of `$rhs` and `cand`, Boogie can assume the existence of a value that satisfies `$Is(_, W)`. This is why Boogie can prove the `assert` above when the Dafny program included the `if ... { var cand := ...; }` statement.

To fix the unsoundness, I did two things.

First, I removed the `where` clause of `$rhs` altogether (except in some cases when `$rhs` was used in the translation of Dafny's `... := *;` assignments, as I described in other comments on this PR). But even if you remove the `where` clause of `$rhs` in my Boogie sketch above, there's still a problem with the `where` clause for `cand`. How come `cand` has a bad `where` clause?

A variable `y` whose type is a possibly empty type (and also a non-ghost variable `y` whose type is not an auto-init type) is subject to definite-assignment rules. The verifier enforces definite-assignment rules using `defass$` variables, as I sketched above. Previously, the verifier implemented an "optimization": a variable declaration with an initial value, like `var y := E;`, was deemed to be "trivially definitely-assigned", and thus the `defass$` machinery was omitted. But this is wrong, because the Boogie program gets to see the `where` clause _before_ the initializing assignment happens (as you can see in the Boogie program above).

So, the second thing I did to fix the unsoundness was to not do this "optimization" for possibly empty types.

After my PR, the Boogie translation of the same method is:

```
procedure M(W: Ty, s: Seq Box where ...) returns (defass$r: bool, r: Box where defass$r ==> $Is(r, W))
{
  var b: bool;
  var defass$w: bool;
  var w: Box where defass$w ==> $Is(w, W);
  var $rhs: Box; // no where clause
  var cand: Box; // no where clause

  // ... same as in the procedure above ...
}
```

Reply:
I see, I didn't realise that the where clauses create assume statements. I wasn't actually sure what they did and assumed they translated to assertions.

Is it a good idea to use where clauses to begin with? If you do a call, then the return type of the callee can add an implicit ensures statement, which translates into an assume on the call site, and if you do a havoc, it makes sense to do an assume on the type afterwards, so then what are the remaining use-cases for where clauses?

Reply:
Added comment.